packages/squiggle-lang/src/fr/list.ts

```diff
@@ -426,28 +410,26 @@ export const library = [
         [frNumber, frLambdaNand([2, 3])],
         "Call with either 2 or 3 arguments, not both."
       ),
-      makeDefinition(
-        [
-          frArray(frGeneric("B")),
-          frGeneric("A"),
-          frLambdaTyped([frGeneric("A"), frGeneric("B")], frGeneric("A")),
-        ],
-        frGeneric("A"),
-        ([array, initialValue, lambda], context) =>
-          _reduce(array, initialValue, lambda, context, false)
-      ),
       makeDefinition(
         [
           frArray(frGeneric("B")),
           frGeneric("A"),
           frLambdaTyped(
-            [frGeneric("A"), frGeneric("B"), frNumber],
+            [frGeneric("A"), frGeneric("B"), frOptional(frNumber)],
             frGeneric("A")
           ),
         ],
         frGeneric("A"),
-        ([array, initialValue, lambda], context) =>
-          _reduce(array, initialValue, lambda, context, true)
+        ([array, initialValue, lambda], context) => {
+          const usedOptional = chooseLambdaParamLength([2, 3], lambda) === 3;
```

Comment:
Flagging that this code is quite hard to read/understand. But I don't have great suggestions on how to improve this.

Reply:
I also thought about this, and iterated a bit.